clients/src/main/java/org/apache/kafka/common/record/MemoryRecords.java

```diff
@@ -115,7 +115,10 @@ public long append(long offset, long timestamp, byte[] key, byte[] value) {
      * to accept this single record.
      */
     public boolean hasRoomFor(byte[] key, byte[] value) {
-        return this.writable && this.compressor.numRecordsWritten() == 0 ?
+        if (!this.writable)
+            return false;
+
+        return this.compressor.numRecordsWritten() == 0 ?
```

Comment:
@granthenke,  javadoc contains some info about the method implementation. Reproducing the javadoc below

>  There is an exceptional case when appending a single message whose size is larger than the batch size, the  capacity will be the message size which is larger than the write limit, i.e. the batch size. In this case  the checking should be based on the capacity of the initialized buffer rather than the write limit in order  to accept this single record

>